app/models/smart_proxy.rb

```diff
@@ -89,15 +89,17 @@ def has_feature?(feature_name)
   end
 
   def capabilities(feature)
-    smart_proxy_features.find_by(:feature_id => Feature.find_by(:name => feature)).try(:capabilities)
+    feature_id = Feature.find_by(:name => feature).try(:id)
```

Comment:
For my understanding, we have `feature_ids = Feature.where(:name => feature_name).pluck(:id)` in `has_feature?`. It feels like this should be the same. Effectively we have 3 places we convert a feature name to an ID very close together, perhaps there are more.

Reply:
has_feature? actually supports multiple features being passed in, which these two methods shouldn't.   Although i'm not sure that i see has_feature? actually being used that way

Reply:
I doubt it - the name is singular so as an API user I wouldn't expect it to work that way.

Reply:
it is not saying explicitly, but if you look at the [implementation](https://github.com/theforeman/foreman/blob/827ee5926edb24dacac13f2413c4359865813953/app/models/smart_proxy.rb#L86-L89) we simply pass the argument to AR, which also accepts an array. So `has_feature?` can be used with one or more feature names.

It may be worth introducing a third method, `find_features_by_name` and `capabilities` method would use it for a single feature. `has_feature?` would only add `pluck` and kept the `any?` part.

It's a small DRY-up I wouldn't insist on, but a valid enhancement.

Reply:
> it is not saying explicitly, but if you look at the [implementation](https://github.com/theforeman/foreman/blob/827ee5926edb24dacac13f2413c4359865813953/app/models/smart_proxy.rb#L86-L89) we simply pass the argument to AR, which also accepts an array. So `has_feature?` can be used with one or more feature names.

As an API consumer I would expect `has_features?` for an array and wouldn't even try passing an array to `has_feature?`.

On the other hand, there's probably more refactoring we can do. For example, I wouldn't mind splitting off all the feature related parts into its own file. `associate_features` is a large method that maybe should even be a service rather than a method on a model. Given that's a larger refactor, I don't insist on DRY-ing up this part.

Writing this down is sometimes more to structure my own thoughts and take others on the same journey than actually asking for it. Then they can make a more informed decision.

Reply:
dry'd up a bit, i can combine with has_feature? if desired.   I don't see any issue within foreman or katello with changing that method to only take 1 in, but its possible some other plugin is doing something silly